crates/turborepo-fs/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+use turbopath::{AbsoluteSystemPathBuf, AnchoredSystemPathBuf};
+use walkdir::WalkDir;
+
+pub fn recursive_copy(src: &AbsoluteSystemPathBuf, dst: &AbsoluteSystemPathBuf) -> Result<()> {
+    let src_metadata = src.metadata()?;
+    if src_metadata.is_dir() {
+        let walker = WalkDir::new(src.as_path()).follow_links(false);
+        for entry in walker.into_iter() {
+            match entry {
+                Err(e) => {
+                    if let Some(_) = e.io_error() {
+                        // Matches go behavior where we translate path errors
+                        // into skipping the path we're currently walking
+                        continue;
+                    } else {
+                        return Err(e.into());
+                    }
+                }
+                Ok(entry) => {
+                    let path = AbsoluteSystemPathBuf::new(entry.path())?;
+                    let file_type = entry.file_type();
+                    // currently we support symlinked files, but not symlinked directories:
+                    // For copying, we Mkdir and bail if we encounter a symlink to a directoy
+                    // For finding packages, we enumerate the symlink, but don't follow inside
+                    // Note that we also don't currently copy broken symlinks
+                    let is_dir_or_symlink_to_dir = if file_type.is_dir() {
+                        true
+                    } else if file_type.is_symlink() {
+                        if let Ok(metadata) = path.stat() {
+                            metadata.is_dir()
+                        } else {
+                            // If we have a broken link, skip this entry
+                            continue;
+                        }
+                    } else {
+                        false
+                    };
+
+                    let suffix = AnchoredSystemPathBuf::new(src, &path)?;
+                    let target = dst.resolve(&suffix);
+                    if is_dir_or_symlink_to_dir {
+                        let src_metadata = entry.metadata()?;
+                        make_dir_copy(&target, &src_metadata)?;
+                    } else {
+                        copy_file_with_type(&path, file_type, &target)?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    } else {
+        copy_file_with_type(src, src_metadata.file_type(), dst)
+    }
+}
+
+fn make_dir_copy(dir: &AbsoluteSystemPathBuf, src_metadata: &Metadata) -> Result<()> {
+    let mut builder = DirBuilder::new();
+    #[cfg(not(windows))]
+    {
+        use std::os::unix::prelude::MetadataExt;
+        builder.mode(src_metadata.mode());
+    }
+    builder.recursive(true);
+    builder.create(dir.as_path())?;
+    Ok(())
+}
+
+pub fn copy_file(from: &AbsoluteSystemPathBuf, to: &AbsoluteSystemPathBuf) -> Result<()> {
+    let metadata = from.metadata()?;
+    copy_file_with_type(from, metadata.file_type(), to)
+}
+
+fn copy_file_with_type(
```

Comment:
As long as we are going this primarily via FFI I think doing this strictly blocking is OK, but once we begin to use this on the rust side, it would be nice if we make the copy calls async so that we can yield the thread.

Reply:
That said, more of our files are probably very small, so maybe we can just leave it sync and throw it in a thread if needs-be.

Reply:
I'm happy to move this to async post-port. For now I'm trying to maintain the existing behavior.